### PR TITLE
Call lower_const_param instead of duplicating the code

### PR DESCRIPTION
**Pull request by spastorino**
Follow up of #123689

r? @oli-obk

I had this commit in my old branch that I had forgotten about, @fmease pointed about this in #123689

I've left the branches that are not `Range` as do nothing as that's what we are currently doing but maybe we want to err or something.

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs [hunk 1]
```diff
@@ -717,6 +717,35 @@ impl<'a, 'tcx> Visitor<'tcx> for BoundVarContext<'a, 'tcx> {
...
+            | hir::PatKind::Never
+            | hir::PatKind::Path(_)
+            | hir::PatKind::Tuple(_, _)
+            | hir::PatKind::Box(_)
+            | hir::PatKind::Deref(_)
+            | hir::PatKind::Ref(_, _)
+            | hir::PatKind::Lit(_)
+            | hir::PatKind::Slice(_, _, _)
+            | hir::PatKind::Err(_) => {}
+        }
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:746]: can you just use `intravisit::walk_pat(p)`?

**Review by compiler-errors — APPROVED** [review]
r=me after fixing the visitor to be simpler

File: compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs [hunk 2]
```diff
@@ -717,6 +717,21 @@ impl<'a, 'tcx> Visitor<'tcx> for BoundVarContext<'a, 'tcx> {
...
+        if let hir::PatKind::Range(start, end, _) = p.kind {
+            if let Some(start) = start {
+                self.visit_expr(start);
+            }
+
+            if let Some(end) = end {
+                self.visit_expr(end);
+            }
+        } else {
+            intravisit::walk_pat(self, p)
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs:731]: Doesn't walk_pat already visit the nested expressions?

Comment (spastorino) [reply compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs]: Right, done!

**Pull request merged**